File: A1_comparing_languages/C#/trees.cs
using System;
using System.Collections.Generic;

public class Trees {
    public static List<string> Gen(int n) {
        List<List<string>> result = new List<List<string>>();

        List<string> tmp = new List<string>();
        tmp.Add("");
        result.Add(tmp);

        if (n > 0) {

            tmp = new List<string>();
            tmp.Add("(.)");
            result.Add(tmp);

            for (int i = 2; i < n+1; i++) {

                tmp = new List<string>();
                for (int j = 0; j < i; j++) {
                    foreach (string left in result[j]) {
                        foreach (string right in result[i-j-1]) {
                            tmp.Add("(" + left + "." + right + ")");
                        }
                    }
                }

                result.Add(tmp);
            }
        }

        return result[result.Count - 1];
    }


    public static void Main(string[] args) {
        int n = Int32.Parse(args[0]);
        List<string> trees = Gen(n);
        foreach (string tree in trees) {
            Console.WriteLine(tree);
        }
    }
}

File: A1_comparing_languages/Python/trees.py
import sys

def gen(n):
    result = []

    # When N is 0
    tmp = ['']
    result.append(tmp)

    if n > 0:
        # When N is 1
        tmp = ['(.)']
        result.append(tmp)

        # When N is greater than 1
        for i in range(2,n+1):
            tmp = []
            # For all possible leftsubtree with j nodes and rightsubtree with i-j-1 nodes
            for j in reversed(range(i)):
                for left in result[j]:
                    for right in result[i-j-1]:
                        # Add leftsubtree and rightsubtree to root and store in list
                        tmp.append('(' + left + '.' + right + ')')
            result.append(tmp)
        
    return result[-1]

def main():
    n = int(sys.argv[1])

    trees = gen(n)
    for tree in trees:
        print(tree)

if __name__ == "__main__":
    main()